Kudu java client bug: NonRecoverableException Scanner xx not found (it may have expired)

presto version: 345

This is a known bug which is unresolved in kudu java client: 
[https://issues.apache.org/jira/browse/KUDU-3169](https://issues.apache.org/jira/browse/KUDU-3169)

I track the stack trace as below
`at io.prestosql.plugin.kudu.KuduRecordCursor.advanceNextPosition(KuduRecordCursor.java:118)`
and found that every time KuduRecordSet will invoke cursor() method, and do  'clientSession.createScanner(kuduSplit)' every time.

Is this possbile that KuduSplit lead the kudu java client send a request to a wrong tserver?


```
Caused by: java.lang.RuntimeException: org.apache.kudu.client.NonRecoverableException: Scanner 2a65adaadd6b4c3da5154ed11b52bbed not found (it may have expired)
	at io.prestosql.plugin.kudu.KuduRecordCursor.advanceNextPosition(KuduRecordCursor.java:124)
	at io.prestosql.$gen.CursorProcessor_20210310_160748_16247.process(Unknown Source)
	at io.prestosql.operator.ScanFilterAndProjectOperator$RecordCursorToPages.process(ScanFilterAndProjectOperator.java:323)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
	at io.prestosql.operator.WorkProcessorUtils$YieldingProcess.process(WorkProcessorUtils.java:181)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
	at io.prestosql.operator.WorkProcessorUtils.lambda$processStateMonitor$2(WorkProcessorUtils.java:200)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils.lambda$flatten$6(WorkProcessorUtils.java:277)
	at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:319)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils$3.process(WorkProcessorUtils.java:306)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
	at io.prestosql.operator.WorkProcessorUtils.lambda$processStateMonitor$2(WorkProcessorUtils.java:200)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorUtils.getNextState(WorkProcessorUtils.java:221)
	at io.prestosql.operator.WorkProcessorUtils.lambda$finishWhen$3(WorkProcessorUtils.java:215)
	at io.prestosql.operator.WorkProcessorUtils$ProcessWorkProcessor.process(WorkProcessorUtils.java:372)
	at io.prestosql.operator.WorkProcessorSourceOperatorAdapter.getOutput(WorkProcessorSourceOperatorAdapter.java:149)
	at io.prestosql.operator.Driver.processInternal(Driver.java:379)
	at io.prestosql.operator.Driver.lambda$processFor$8(Driver.java:283)
	at io.prestosql.operator.Driver.tryWithLock(Driver.java:675)
	at io.prestosql.operator.Driver.processFor(Driver.java:276)
	at io.prestosql.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1076)
	at io.prestosql.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
	at io.prestosql.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:484)
	at io.prestosql.$gen.Presto_345____20210310_104032_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.apache.kudu.client.NonRecoverableException: Scanner 2a65adaadd6b4c3da5154ed11b52bbed not found (it may have expired)
	at org.apache.kudu.client.KuduException.transformException(KuduException.java:110)
	at org.apache.kudu.client.KuduClient.joinAndHandleException(KuduClient.java:413)
	at org.apache.kudu.client.KuduScanner.nextRows(KuduScanner.java:72)
	at io.prestosql.plugin.kudu.KuduRecordCursor.advanceNextPosition(KuduRecordCursor.java:118)
	... 32 more
	Suppressed: org.apache.kudu.client.KuduException.OriginalException: Original asynchronous stack trace
		at org.apache.kudu.client.RpcProxy.dispatchTSError(RpcProxy.java:347)
		at org.apache.kudu.client.RpcProxy.responseReceived(RpcProxy.java:269)
		at org.apache.kudu.client.RpcProxy.access$000(RpcProxy.java:59)
		at org.apache.kudu.client.RpcProxy$1.call(RpcProxy.java:149)
		at org.apache.kudu.client.RpcProxy$1.call(RpcProxy.java:145)
		at org.apache.kudu.client.Connection.messageReceived(Connection.java:390)
		at org.apache.kudu.shaded.org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
		at org.apache.kudu.client.Connection.handleUpstream(Connection.java:238)
		at org.apache.kudu.shaded.org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
		at org.apache.kudu.shaded.org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
		at org.apache.kudu.shaded.org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
		at org.apache.kudu.shaded.org.jboss.netty.handler.codec.oneone.OneToOneDecoder.handleUpstream(OneToOneDecoder.java:70)
		at org.apache.kudu.shaded.org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
		at org.apache.kudu.shaded.org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
		at org.apache.kudu.shaded.org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
		at org.apache.kudu.shaded.org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
		at org.apache.kudu.shaded.org.jboss.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
		at org.apache.kudu.shaded.org.jboss.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303)
		at org.apache.kudu.shaded.org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
		at org.apache.kudu.shaded.org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
		at org.apache.kudu.shaded.org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
		at org.apache.kudu.shaded.org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
		at org.apache.kudu.shaded.org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
		at org.apache.kudu.shaded.org.jboss.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
		at org.apache.kudu.shaded.org.jboss.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303)
		at org.apache.kudu.shaded.org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
		at org.apache.kudu.shaded.org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
		at org.apache.kudu.shaded.org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
		at org.apache.kudu.shaded.org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
		at org.apache.kudu.shaded.org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
		at org.apache.kudu.shaded.org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
		at org.apache.kudu.shaded.org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
		at org.apache.kudu.shaded.org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:337)
		at org.apache.kudu.shaded.org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
		at org.apache.kudu.shaded.org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
		at org.apache.kudu.shaded.org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
		at org.apache.kudu.shaded.org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
		... 3 more
```

--- a/plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduRecordCursor.java
+++ b/plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduRecordCursor.java
@@ -23,10 +23,10 @@ import io.trino.spi.type.Type;
 import org.apache.kudu.client.KeyEncoderAccessor;
 import org.apache.kudu.client.KuduException;
 import org.apache.kudu.client.KuduScanner;
+import org.apache.kudu.client.KuduScannerIterator;
 import org.apache.kudu.client.KuduTable;
 import org.apache.kudu.client.PartialRow;
 import org.apache.kudu.client.RowResult;
-import org.apache.kudu.client.RowResultIterator;
 
 import java.lang.reflect.Field;
 import java.util.List;
@@ -55,11 +55,10 @@ public class KuduRecordCursor
     private final List<Type> columnTypes;
     private final KuduTable table;
     private final Map<Integer, Integer> fieldMapping;
-    private RowResultIterator nextRows;
+    private final KuduScannerIterator kuduScannerIterator;
     private RowResult currentRow;
 
     private long totalBytes;
-    private boolean started;
 
     public KuduRecordCursor(KuduScanner scanner, KuduTable table, List<Type> columnTypes, Map<Integer, Integer> fieldMapping)
     {
@@ -67,6 +66,7 @@ public class KuduRecordCursor
         this.columnTypes = ImmutableList.copyOf(requireNonNull(columnTypes, "columnTypes is null"));
         this.table = requireNonNull(table, "table is null");
         this.fieldMapping = ImmutableMap.copyOf(requireNonNull(fieldMapping, "fieldMapping is null"));
+        kuduScannerIterator = scanner.iterator();
     }
 
     @Override
@@ -98,30 +98,11 @@ public class KuduRecordCursor
     @Override
     public boolean advanceNextPosition()
     {
-        boolean needNextRows = !started || !nextRows.hasNext();
-
-        if (!started) {
-            started = true;
+        if (!kuduScannerIterator.hasNext()) {
+            return false;
         }
 
-        if (needNextRows) {
-            currentRow = null;
-            try {
-                do {
-                    if (!scanner.hasMoreRows()) {
-                        return false;
-                    }
-
-                    nextRows = scanner.nextRows();
-                }
-                while (!nextRows.hasNext());
-            }
-            catch (KuduException e) {
-                throw new RuntimeException(e);
-            }
-        }
-
-        currentRow = nextRows.next();
+        currentRow = kuduScannerIterator.next();
         totalBytes += getRowLength();
         return true;
     }
@@ -204,9 +185,6 @@ public class KuduRecordCursor
     @Override
     public void close()
     {
-        if (!started) {
-            return;
-        }
         try {
             scanner.close();
         }
@@ -214,6 +192,5 @@ public class KuduRecordCursor
             throw new RuntimeException(e);
         }
         currentRow = null;
-        nextRows = null;
     }
 }

--- a/plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduRecordCursor.java
+++ b/plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/KuduRecordCursor.java
@@ -15,7 +15,6 @@ package io.trino.plugin.kudu;
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
-import io.airlift.log.Logger;
 import io.airlift.slice.Slice;
 import io.airlift.slice.Slices;
 import io.trino.spi.TrinoException;
@@ -40,8 +39,6 @@ import static java.util.Objects.requireNonNull;
 public class KuduRecordCursor
         implements RecordCursor
 {
-    private static final Logger log = Logger.get(KuduRecordCursor.class);
-
     private static final Field ROW_DATA_FIELD;
 
     static {
@@ -118,7 +115,6 @@ public class KuduRecordCursor
                     nextRows = scanner.nextRows();
                 }
                 while (!nextRows.hasNext());
-                log.debug("Fetched " + nextRows.getNumRows() + " rows");
             }
             catch (KuduException e) {
                 throw new RuntimeException(e);

